src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs

```diff
@@ -392,5 +805,89 @@ private static float L2DistSquared(Span<float> a, Span<float> b)
                 return norm;
             }
         }
+
+        public static void ZeroMatrixItems(AlignedArray dst, int ccol, int cfltRow, int[] indices)
+        {
+            Contracts.Assert(0 < ccol && ccol <= cfltRow);
+
+            // REVIEW NEEDED: Since the two methods below do not involve any SSE hardware intrinsics, no software fallback is needed.
```

Comment:
I'm not understanding here.  If there are no SIMD operations involved in these methods, then let's just have the software function in `CpuMathUtils` and not have the method at all in `SseIntrinsics`.  `SseIntrinsics` should be only for when we use the `Sse` operations.

Reply:
`ZeroItemsU` and `ZeroMatrixItemsCore` are two functions in `Sse.cpp` like all other functions currently in `SseIntrinsics.cs`, but since even their native implementation does not involve Intel hardware intrinsics like `_mm_set_ps` at all, there is no need to implement software fallbacks.  We require SSE support only for us to call those Intel hardware intrinsics, which are absent in `ZeroItemsU` and `ZeroMatrixItemsCore`.

While the implementations of these two functions are software-based, they do involve pointers and follow the C++ syntax in `Sse.cpp`.  I can put these two functions in `CpuMathUtils` and rewrite them so that they follow the C# syntax.

Reply:
Before I proceed to modify these function, may I know if I have given relevant info, and if this change is still desired?

Reply:
I think the right change here is just to move your current code (using pointers) from `SseIntrinsics` into `CpuMathUtils`.  I assume the reason it is using pointers is for speed.  Let's keep it that way until it can be proved we don't need pointers.

Reply:
Agree - thanks!  `ZeroItemsU` and `ZeroMatrixItemsCore` call the `Ptr` method which is private in `SseIntrinsics`.  Should I make `Ptr` in `SseIntrinsics` `internal` instead of `Private`?

The above change could solve the issue.  Another way is to remove the dependence on `Ptr` in `CpuMathUtils`:
```
        private static unsafe void ZeroItemsU(AlignedArray dst, int c, int[] indices, int cindices)
        {
            for (int i = 0; i < cindices; i++)
            {
                int index = indices[i];
                Contracts.Assert(index >= 0);
                Contracts.Assert(index < c);
                dst[index] = 0;
            }

            /* Original version calls the Ptr method, which is private in SseIntrinsics
            fixed (float* pDstStart = &dst.Items[0])
            fixed (int* pidx = &indices[0])
            {
                float* pdst = Ptr(dst, pDstStart);

                for (int i = 0; i < cindices; ++i)
                {
                    int index = pidx[i];
                    Contracts.Assert(0 <= index && index < c);
                    pdst[index] = 0;
                }
            }
            */
        }

        private static unsafe void ZeroMatrixItemsCore(AlignedArray dst, int c, int ccol, int cfltRow, int[] indices, int cindices)
        {
            int ivLogMin = 0;
            int ivLogLim = ccol;
            int ivPhyMin = 0;

            for (int i = 0; i < cindices; i++)
            {
                int index = indices[i];
                Contracts.Assert(index >= 0);
                Contracts.Assert(index < c);

                int col = index - ivLogMin;
                if ((uint)col >= (uint)ccol)
                {
                    Contracts.Assert(index < ivLogMin || index >= ivLogLim);

                    int row = index / ccol;
                    ivLogMin = row * ccol;
                    ivLogLim = ivLogMin + ccol;
                    ivPhyMin = row * cfltRow;

                    Contracts.Assert(index >= ivLogMin);
                    Contracts.Assert(index < ivLogLim);
                    col = index - ivLogMin;
                }

                dst[ivPhyMin + col] = 0;
            }

            /* Original version calls the Ptr method, which is private in SseIntrinsics
            fixed (float* pDstStart = &dst.Items[0])
            fixed (int* pidx = &indices[0])
            {
                float* pdst = Ptr(dst, pDstStart);

                int ivLogMin = 0;
                int ivLogLim = ccol;
                int ivPhyMin = 0;

                for (int i = 0; i < cindices; ++i)
                {
                    int index = pidx[i];
                    Contracts.Assert(0 <= index && index < c);

                    int col = index - ivLogMin;
                    if ((uint)col >= (uint)ccol)
                    {
                        Contracts.Assert(ivLogMin > index || index >= ivLogLim);

                        int row = index / ccol;
                        ivLogMin = row * ccol;
                        ivLogLim = ivLogMin + ccol;
                        ivPhyMin = row * cfltRow;

                        Contracts.Assert(ivLogMin <= index && index < ivLogLim);
                        col = index - ivLogMin;
                    }

                    pdst[ivPhyMin + col] = 0;
                }
            }
            */
        }
```
 Please let me know which approach I should follow (the former or latter).

Reply:
You're "latter" approach isn't using pointers at all.

I was thinking we wouldn't need to call `Ptr` at all.  That method moves the array around in memory so it is guaranteed to be aligned.  Since we aren't using SIMD instructions to set all the values to `0`, we can just pin the arrays using

```
fixed (float* pDstStart = &dst.Items[0])
fixed (int* pidx = &indices[0])
```

And then continue with the logic without calling `Ptr`.

Reply:
Thank you!  I think this is doable, but `pdst[index] = 0` in `ZeroItemsU` and `pdst[ivPhyMin + col] = 0` will be affected if we do not call `Ptr` and increment `pDstStart` to get the pointer `pdst` that points to the base of `dst`.  How should I proceed?

Reply:
You can just use the pointer you have when you pinned the array:

```C#
        internal static unsafe void ZeroItemsU(AlignedArray dst, int c, int[] indices, int cindices)
        {
            fixed (float* pdst = &dst.Items[0])
            fixed (int* pidx = &indices[0])
            {
                for (int i = 0; i < cindices; ++i)
                {
                    int index = pidx[i];
                    Contracts.Assert(0 <= index && index < c);
                    pdst[index] = 0;
                }
            }
        }
```